### PR TITLE
refactor(i18n): expose hook for AsyncLocaleData

### PR DESCRIPTION
**Pull request by emmenko**
- expose a `useAsyncLocaleData` hook. The `<AsyncLocaleData>` component now uses the hook internally.
- expose a `useAsyncIntlMessages` hook, to only load messages with dynamic imports (code splitting).

The second hook is useful for building UI packages (not custom apps) that ship with translations. The translation message should be encapsulated with the package itself and ideally they should be code splitted as well.
The UI package then can use a `IntlProvider` (which will be nested when used within a Custom Application) to load the messages of the package itself.
The `IntlProvider` then merges the messages with the "parent(s)" providers.

### CODE REVIEW

File: packages/i18n/README.md [hunk 1]
```diff
@@ -20,21 +20,27 @@ $ npm install --save @commercetools-frontend/i18n
...
+> The `<AsyncLocaleData>` component, or the `useAsyncLocaleData` hook, are internally used by the `<ApplicationShell>` and there is no need to use them directly.
+> If you do need to load translation files asynchronously, we expose an additional hook `useAsyncIntlMessages` that you can use for that.
+> The difference between those is that the "async locale data" components additionally load the application-kit and ui-kit messages, as well as the moment locales.
+
+### Using a React component with render props
+
+**With static messages**
 
 ```js
+import { IntlProvider } from 'react-intl';
```
Comment (tdeekens) [inline packages/i18n/README.md:34]: This change is unrelated?

File: packages/i18n/src/async-locale-data/async-locale-data.tsx [hunk 2]
```diff
@@ -32,63 +34,56 @@ const getMessagesForLocale = (data?: TMessages, locale?: string) => {
...
+  const loadApplicationMessages = React.useCallback(
+    async (locale: string) => {
+      if (typeof applicationMessages === 'function') {
+        return await applicationMessages(locale);
+      }
+      return getMessagesForLocale(applicationMessages, locale);
+    },
+    [applicationMessages]
+  );
+  const messagesFromKitResult = useAsyncIntlMessages({
```
Comment (tdeekens) [inline packages/i18n/src/async-locale-data/async-locale-data.tsx:50]: Kit is AppKit?

File: packages/i18n/src/async-locale-data/use-async-intl-messages.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  error: undefined,
+};
+
+// Low level hook to load messages for a specific locale. The loading is async
+// because it's assumed that the translation files are dynamically imported (code splitted).
+const useAsyncIntlMessages = ({ locale, loader }: THookOptions): TState => {
+  const [state, setState] = React.useState(initialState);
+
+  React.useEffect(() => {
+    let _isUnmounting = false;
```
Comment (tdeekens) [inline packages/i18n/src/async-locale-data/use-async-intl-messages.ts:29]: I think I do the same in flopflip and use a ref inside a hook. Like `const isUnmountedRef = useIsUnmounted()`. Can't recall I just remember it took longer than I thought to get right. I'll check later.

Comment (emmenko) [reply packages/i18n/src/async-locale-data/use-async-intl-messages.ts:29]: I took this approach from reading this: https://stackoverflow.com/a/56443045

File: packages/i18n/src/async-locale-data/async-locale-data.tsx (hunk 2 above)
Comment (emmenko) [reply packages/i18n/src/async-locale-data/async-locale-data.tsx:50]: Both, app and ui kit

File: packages/i18n/README.md (hunk 1 above)
Comment (emmenko) [reply packages/i18n/README.md:34]: Yes and now. Since I'm updating the README though, to avoid confusion I think it would be better to keep the example generic and use `IntlProvider`, not our internal `ConfigureIntlProvider`.

**Comment by emmenko** [comment]
FYI: not a high prio to review. Whenever you have time...

**Comment by adnasa** [comment]
I can't review this today, sorry. I'll have a look at this next week.

File: packages/application-shell/src/components/project-suspended/project-suspended.spec.js [hunk 4]
```diff
@@ -98,12 +98,9 @@ describe('rendering', () => {
...
-      expect(
-        rendered.getByText(
-          /Your Project is temporarily suspended due to maintenance/
-        )
-      ).toBeInTheDocument();
-    });
+    await screen.findByText(
+      /Your Project is temporarily suspended due to maintenance/
+    );
+    await screen.findByText('Search for a project');
```
Comment (adnasa) [inline packages/application-shell/src/components/project-suspended/project-suspended.spec.js:104]: I'm liking `screen` alot

File: packages/i18n/README.md [hunk 5]
```diff
@@ -100,28 +106,51 @@ const Application = (props) => (
...
+Similarly to the `<AsyncLocaleData>`, we also expose a React hook.
+
+```js
+import { IntlProvider } from 'react-intl';
+import { useAsyncLocaleData } from '@commercetools-frontend/i18n';
+
+const Application = (props) => {
+  const { isLoading, locale, messages } = useAsyncLocaleData({
+    locale: props.locale,
+    applicationMessages: loadMessages,
```
Comment (adnasa) [inline packages/i18n/README.md:129]: where is `loadMessages` coming from?

**Comment by tdeekens** [comment]
@emmenko should we proceed here?

File: packages/i18n/README.md (hunk 5 above)
Comment (emmenko) [reply packages/i18n/README.md:129]: It's defined in the example above. Didn't want to copy-paste the function again.

**Comment by emmenko** [comment]
Yep, we can merge this.
